### ISSUE TITLE
[Feature]: Update `psr/log` version or make it less strict

### ISSUE THREAD

**Issue by hirasso** [snicco/snicco#204]
### Thanks for taking the time to fill out this feature request!"

- [X] I have searched the issue tracker for similar requests

### Which @snicco/* package(s) are relevant/related of the feature request?

http-routing

### Description

I can't install the **[HttpRoutingBundle](https://github.com/snicco/http-routing-bundle)** due to version conflicts in my project:

```
❯ composer require snicco/http-routing-bundle

Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - snicco/http-routing-bundle[dev-master, v1.0.0, ..., v1.10.0] require psr/log ^1.1.1 -> found psr/log[1.1.1, 1.1.2, 1.1.3, 1.1.4] but the package is fixed to 3.0.0 (lock file version) by a partial update and that version does not match. Make sure you list it as an argument for the update command.
    - Root composer.json requires snicco/http-routing-bundle * -> satisfiable by snicco/http-routing-bundle[dev-master, v1.0.0, ..., v1.10.0, 9999999-dev].
```

These are the packages that I already have in use in my project that make use of `psr/log`:
```
❯ composer why psr/log
composer/xdebug-handler 3.0.5 requires psr/log (^1 || ^2 || ^3) 
monolog/monolog         3.7.0 requires psr/log (^2.0 || ^3.0)  
```

### Proposed solution

Maybe it's possible to make the version of the [`psr/log` dependency](https://github.com/snicco/http-routing-bundle/blob/aabfd573bca8e58929817315731d5a40e9d92bb1/composer.json#L21) less strict, as in the packages above?

### Alternatives

I don't know of any – does anyone else?

**Comment by calvinalkan** [comment]
I don't remember exactly why, but the 2.0.0 release broke a bunch of tests, so I hit with the hammer and locked it to 1.X 

I think it was because some test helpers were removed. 

I think it might be possible to have 1.X as require dev, and X.X as require.

**Comment by hirasso** [comment]
Any chance this might be fixed anytime soon? Otherwise, I'll need to hit with the hammer and scope my plugin even for installs via composer 😅

**Comment by hirasso** [comment]
...removing the claim "zero dependencies" from all the readmes would probably also be a good thing...

**Comment by calvinalkan** [comment]
> ...removing the claim "zero dependencies" from all the readmes would probably also be a good thing...

Many packages have some deps, many don't.

I'm confused, are you trying to install `snicco/snicco`?

You should just install what you need, for example:

`composer require snicco/better-wpdb` 

which has no dependencies:

[snicco/better-wpdb](https://github.com/snicco/better-wpdb/blob/a29c0f0bfaf6e104537f1678e1a64a82d047fa41/composer.json#L10)

**Comment by hirasso** [comment]
I'm currently trying to install snicco/better-wp-cli. In the same project I'm using monolog/logger@3, which requires psr/log@3. better-wp-cli requires psr/log@1 and thus can't be installed.

**Comment by hirasso** [comment]
Maybe this is all just a matter of declaring the correct dependencies in the snicco sub-packages?

**Comment by hirasso** [comment]
In better-wp-cli, psr/log is required here:

https://github.com/snicco/better-wp-cli/blob/c52e08ab6ecb9574aa3445e6e0915c3077b45805/composer.json#L13

And it's also being used by: 

https://github.com/snicco/better-wp-cli/blob/c52e08ab6ecb9574aa3445e6e0915c3077b45805/src/Logger/PSRLogger.php#L7

**Comment by calvinalkan** [comment]
> Maybe this is all just a matter of declaring the correct dependencies in the snicco sub-packages?

Yes, the requirements are overly strict, I think.

This is because during **testing** we use the PSR `TestLogger` which was removed in version 2 of the `psr/log`.

In production, any version of psr log should work fine.

I'll see if composer supports different package constraints during development vs prod.

**Comment by calvinalkan** [comment]
> In better-wp-cli, psr/log is required here:
> 
> https://github.com/snicco/better-wp-cli/blob/c52e08ab6ecb9574aa3445e6e0915c3077b45805/composer.json#L13
> 
> And it's also being used by:
> 
> https://github.com/snicco/better-wp-cli/blob/c52e08ab6ecb9574aa3445e6e0915c3077b45805/src/Logger/PSRLogger.php#L7

You're right. I probably forgot to update the readme when I added psr/log.

**Comment by calvinalkan** [comment]
> Any chance this might be fixed anytime soon? Otherwise, I'll need to hit with the hammer and scope my plugin even for installs via composer 😅

You should always scope your plugin if you distribute it to third-parties.

If you don't distribute it, you don't need to scope.
